[Encryption] New component

Nyholm opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.0
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | 
| License       | MIT
| Doc PR        | 

This PR introduces a new Encryption component. But wait, NO. That is not what we are doing. 

### We are not doing encryption

This component does not do encryption. We do not invent a way to protect your data. This is an **encryption abstraction component**. So the only thing we do is to define two interfaces and provide a sodium implementation of them. 

### The interfaces

There is an [KeyInterface](https://github.com/Nyholm/symfony/blob/encryption/src/Symfony/Component/Encryption/KeyInterface.php) and a [EncryptionInterface](https://github.com/Nyholm/symfony/blob/encryption/src/Symfony/Component/Encryption/EncryptionInterface.php). They look something like this: 

```php
interface KeyInterface
{
    public function extractPublicKey(): self;
}
```
```php
interface EncryptionInterface
{
    public function generateKey(string $secret = null): KeyInterface;
    public function encrypt(string $message, KeyInterface $myKey): string;
    public function encryptFor(string $message, KeyInterface $recipientKey): string;
    public function decrypt(string $message, KeyInterface $key, KeyInterface $senderPublicKey = null): string;
}
```

There are some documentation on the interfaces, they can be good to read if one is not sure what the methods are doing. 

### Keys
The idea behind these interfaces is that you create a `Key`. That key can be used for both symmetric and asymmetric encryption. That key can be serialized and stored on disk or in database. You can extract the public part of your `Key`.

You as a user need to know what type of key to have. You cannot load a random key from disk and use it with `encryptFor()`. You need to know that it was Alice's private key. Ie, Key management is out of scope of the component. 

Keys cannot be transfered to a different implementation. (Just like PSR-6). So you cannot use `SodiumEncryption` to create a Key and then use it with `PhpseclibEncryption`.

### Encryption

The encryption interface provides an explicit way how to handle encryption. There are 3 scenarios: 
- **encrypt()**: When you don't share the encrypt/decrypt keys
- **encryptFor()**: When you want to send a message to Bob

For someone who worked with encryption they know that this is no rocket science. But for someone who takes their first step into encrypting in PHP, they probably look at [PHP's Sodium "documentation"](https://www.php.net/manual/en/ref.sodium.php) and have no idea where to start. This `EncryptionInterface` will be super helpful for them. 

### Portability

( This section is where I accept all kinds of "you are doing it wrong" 😄 ) 

The output of all `EncryptionInterface::encrypt*()` methods is a string. That string has a specific format. 
The format is implemented in the `Ciphertext` class. Here is how the final string look like: 

```
{"alg": "sodium"} . encryption-output . hash
```

With some psudo-code: 

```php
$nonce = random_bytes(16);
$ciphertext = sodium_encrypt('my message', $nonce);
$headers = json_encode(['alg'=>'sodium_encrypt', 'ver'=>'1', 'nonce' => $nonce]);
$hash = hash('sha256', $headers.$ciphertext);

$output = sprintf('%s.%s.%s',
           base64_encode(headers),
           base64_encode($ciphertext),
           base64_encode($hash)
        );
```

But why do we need this format?

We do need **a** format. We need to store some meta data, the ciphertext and the nonce somewhere. To concatenate these together seams like the simplest solution. 

The `alg` header is basically a routing key to tell the difference between “Algorithm not supported” and “Wrong key was used to decrypt message”. Also to make the keys somewhat portable. Ie, it is possible for a non-encryption-component-application to decrypt the key. Note that there is no algorithm negotiation. Note also that the hash here is not to make things "more secure", it is just for an integrity check. 

This format will also fall under or BC policy. 

I have looked at using JWE, but I was not able to follow the specification to 100% to create the Authentication Tag (Im not sure it is possible in PHP...) and I've also seen [posts about](https://paragonie.com/blog/2017/03/jwt-json-web-tokens-is-bad-standard-that-everyone-should-avoid) not to use JWE. JWE also made things much more complex. 

### The name of the component

I am not a big fan of the name of this component. What I've noticed that just because it is called "Encryption", then people think that we are doing advanced security stuff. This component will never compete with Sodium, Phpseclib or Halite. It will only make a very thin abstraction on top of it. This abstraction will work for 80% of your encryption needs. For the other 20% you need more flexibility or customisation. That is where you use Sodium, Phpseclib, Halite etc directly. 

Ie, this component is not a good fit for "That other team's Java application sends us encrypted data, can we decrypt it?".

Nyholm committed:
Adding new Encryption component

Nyholm commented on src/Symfony/Component/Security/Core/Encryption/SymmetricEncryptionInterface.php:
```diff
@@ -0,0 +1,50 @@
+     *
+     * @throws EncryptionException
+     */
+    public function encrypt(string $message): string;
```
It could be strange that we don't define a `$key` argument here. We are forcing all implementations of `SymmetricEncryptionInterface` to accept a key in the constructor.

fabpot replied:
That's a design decision we might discuss, but why not do the same for the asymetric encryption?

Nyholm replied:
Here is how I've been reasoning with myself. I might be wrong so just disagree with me if you want. 

**Symmetric**: This is used when the same application encrypts and decrypts the message. Ie, storing a value on disk, or when you send a user a magic login link, you want to make sure the query parameters has not been tampered with. Any situation where you want to hide a value. Like `sha1()` but also reversible. 

There is no need for you to have multiple keys. It will never be more secure. You are also never sharing the key, there is no such strategy like "if I have multiple keys, then if one get exposed, all my secrets aren't exposed". (Either none of your key is exposed, or all of them)

That is why the `$key` is not a runtime config, but something you configure in build time. 

**Asymmetric**:
This is about encrypting data to send to other computers. It is also about validating signatures. If you want to send data two 5 computers, you don't want to instantiate 5 different objects if `AsymmetricEncryptionInterface`. 

The Asymmetric keys are also likely to be stored in a database/datastore.

fabpot commented on src/Symfony/Component/Security/Core/Encryption/AsymmetricEncryptionInterface.php:
```diff
@@ -0,0 +1,65 @@
+ * When Bob receive the message, he will decrypt it with keypair [ bob_private, alice_public ].
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ * @experimental in 5.2
```
5.3 (and you should add a newline before)

fabpot commented on src/Symfony/Component/Security/Core/Encryption/AsymmetricEncryptionInterface.php:
```diff
@@ -0,0 +1,65 @@
+interface AsymmetricEncryptionInterface
+{
+    /**
+     * Generate new a keypair to be used with encryption. Don't loose your private
```
Generates a new keypair

fabpot replied:
I would make the "Don't..." a new "paragraph"?

fabpot commented on src/Symfony/Component/Security/Core/Encryption/AsymmetricEncryptionInterface.php:
```diff
@@ -0,0 +1,65 @@
+     * Generate new a keypair to be used with encryption. Don't loose your private
+     * key and make sure to keep it a secret.
+     *
+     * @return array{public: string, private: string}
```
We are not using this format anywhere if I'm not wrong

Nyholm replied:
True, and we never define the keys of an array either. The best I can do following our current pattern is to add `array<string,string>`.

I could also use ArrayShape php8 attribute from [PHPStorm](https://packagist.org/packages/jetbrains/phpstorm-attributes) like: `#[ArrayShape([‘ public’ => ‘string’, ‘ private’ => ‘Foo’,])]`. 

Or, define a small object for a `Keypair`. 

I'll go either way.

fabpot commented on src/Symfony/Component/Security/Core/Encryption/AsymmetricEncryptionInterface.php:
```diff
@@ -0,0 +1,65 @@
+    public function generateKeypair(): array;
+
+    /**
+     * When Alice wants to send a secret message to Bob.
```
To be consistent with the way we do it elsewhere, it should be something along the lines of: "Encrypts a message".

fabpot commented on src/Symfony/Component/Security/Core/Encryption/SodiumEncryption.php:
```diff
@@ -0,0 +1,124 @@
+class SodiumEncryption implements SymmetricEncryptionInterface, AsymmetricEncryptionInterface
+{
+    /**
+     * @var string application secret
```
What about moving it to the constrcutor instead?

fabpot commented on src/Symfony/Component/Security/Core/Encryption/SymmetricEncryptionInterface.php:
```diff
@@ -0,0 +1,50 @@
+
+/**
+ * Symmetric encryption uses the same key to encrypt and decrypt a message. The
+ * keys should be kept safe and should not be exposed to the public. Symmetric
```
key instead of keys?

fabpot commented on src/Symfony/Component/Security/Core/Exception/EncryptionException.php:
```diff
@@ -0,0 +1,19 @@
+/**
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+class EncryptionException extends \Exception implements EncryptionExceptionInterface
```
We tend to avoid extending `\Exception` directly, any other more precise exception class we can use? Maybe `\RuntimeException`

Nyholm replied:
Thank you

fabpot reviewed (changes requested):
Looks georgous and simple! Thank you for doing it. Sorry for the not so useful comments, but I've focused on the text instead of the code for this first review.

Two questions that come to mind:

Would it make sense to have a dedicated component instead of having it in security? I can see many usages for which  I would not need anything else from the Security components.

We are using "encryption" in the secret feature (part of framework bundle) and in the Mime component. Any code that could be moved to this new hypothetical encryption component?

Nyholm commented:
Thank you for the review. I've updated the PR accordingly. 

> Would it make sense to have a dedicated component instead of having it in security?

Yes, it will be a small component, but I do think so. 

> We are using "encryption" in the secret feature (part of framework bundle) and in the Mime component. Any code that could be moved to this new hypothetical encryption component?

Looking at the `SodiumVault`: Yes, we can use these abstractions. They authors have used the sodium keypair as "decryption key" which I fail to understand why, But we can discuss that in a separate issue.

In the MimeComponent we have a `DkimSigner`, `SMimeEncryoter` and `SMimeSigner`, these cannot be leveraged by the `SymmetricEncryptionInterface` or `AsymmetricEncryptionInterface`. 

The current implementation is an abstraction`SymmetricEncryptionInterface` or `AsymmetricEncryptionInterface`, you are not able to chose algorithm. The output of `encrypt()` is following the pattern of `cipher.algorithm.nonce`, so it is possible for a non-php application to decrypt it, but they need to know our arbitrary values of algorithm.

fabpot commented:
There is no such thing as a "small component" :) History showed us that most small components grew over time with new needs are discovered.
So, I'm :+1: for a new component.

wouterj commented:
:+1: for a new component (I would even prefer a component outside the `Security` namespace) and moving its framework integration to the FrameworkBundle.

There already are a couple features in the framework bundle that can benefit from this new component I think, e.g. encrypting cache values (refs #35019) and encrypting session data (refs #35804).

jderusse commented on src/Symfony/Component/Security/Core/Encryption/AsymmetricEncryptionInterface.php:
```diff
@@ -0,0 +1,65 @@
+     *
+     * @return array{public: string, private: string}
+     */
+    public function generateKeypair(): array;
```
Would be great to use this new component in SodiumVault.
In such situation we'll also need a method to generate a public key from a given private key.

Nyholm replied:
Yeah. I saw that.. Im not sure why the SodiumVault keep the keypair instead of the private key. If the SodiumVault kept the private key and public key, it would be a perfect fit.

Nyholm replied:
Fixed!

jderusse commented on src/Symfony/Component/Security/Core/Encryption/PhpseclibEncryption.php:
```diff
@@ -0,0 +1,160 @@
+            if (null === $publicKey) {
+                $algorithm = 'aes';
+                $aes = new AES();
+                $aes->setKey($this->secret);
```
Hmm this class is Asymmetric when calling `encrypt` with a `publicKey` arguments, in that case, it does not use the compulsory argument `secret`. Something looks fragile and complicated here.

I'd rather
```
class PhpseclibSymmetricEncryption implement SymmetricEncryptionInterface
class PhpseclibAsymmetricEncryption implement AsymmetricEncryptionInterface
```
or something similar to https://github.com/paragonie/paseto/tree/master/src/Keys
```
class PhpseclibEncryption implement EncryptionInterface
{
    function __construct(KeyInterface $key) {
    } 
}

interface SymmetricKeyInterface extends KeyInterface
interface AsymmetricKeyInterface extends KeyInterface
```

Nyholm replied:
The concept of "secret" does not exist in RSA (only private key). I use AES only for symmetric encryption, RSA is for asymmetric.

Nyholm commented on src/Symfony/Component/Encryption/Ciphertext.php:
```diff
@@ -0,0 +1,87 @@
+
+    public function getUrlSafeRepresentation(): string
+    {
+        return sprintf('%s.%s.%s', base64_encode($this->ciphertext), base64_encode($this->algorithm), base64_encode($this->nonce));
```
I will change this to be a more JWT/JWE like string

Nyholm commented:
I need some more work to support JWE. Using JWE has the benefit that we more easily can share the encrypted strings with other applications. 

JWE exists of 5 parts: 

**header**: a json object defining what algorithms we use
**CEK**: The encoded key for the symmetric encryption
**Initialization Vector**: Nonce for symmetric encryption
**Cipher text**: The encrypted payload
**Authentication tag**: used for integrity checks

So we use symmetric encryption for the payload and asymmetric to encrypt the symmetric key.

Nyholm commented:
So, we are compatible with the JWE format.. almost. I cannot figure out the `JWE Authentication Tag`. 

Here is an example for [RFC7516](https://tools.ietf.org/html/rfc7516#appendix-A.1.6): 

>   Perform authenticated encryption on the plaintext with the AES GCM algorithm using the CEK as the encryption key, the JWE Initialization Vector, and the Additional Authenticated Data value above, requesting a 128-bit Authentication Tag output.

So basically: "Take a bunch of data, encrypt it with AES GCM and from that encryption you get ciphertext X and Authentication Tag Y". 

But when I look at ie [sodium](https://www.php.net/manual/en/function.sodium-crypto-aead-aes256gcm-encrypt.php) or [libsecphp](phpseclib.sourceforge.net), I find no encryption algoritm where I can get two outputs.. Am I missing something?

However, We are following the same protocol and the ciphertext is secure, but it is not according to the spec (not so good for portability)

norkunas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -1888,4 +1890,16 @@ private function addRateLimiterSection(ArrayNodeDefinition $rootNode)
+        $rootNode
+            ->children()
+                ->arrayNode('encryption')
+                    ->info('esi configuration')
```
```suggestion
                    ->info('encryption configuration')
```
? :)

Nyholm replied:
Oops. Thank you

jderusse commented on src/Symfony/Component/Encryption/Provider/PhpseclibEncryption.php:
```diff
@@ -0,0 +1,183 @@
+     */
+    public function __construct(string $secret)
+    {
+        $this->secret = $secret;
```
Sound's weird to have this compulsory parameter which is not needed for asymmetric crytography.

I get your point that symmetric is more probably used with local secret, while asymmetric public key is more probably provided by end-user.
But this is not always the case (see SodiumVault class) and asymmetric private key is more probably kept local.

IMHO the API should be identical for all kind of cryptography. Either by providing a key in the constructor or in the method.

Nyholm replied:
I dont think it is a good idea to provide private/public keys in the constructor for asymmetric encryption. But I get your point. But that means I need to modify the SymmetricEncryptionInterface to accept a second parameter on the `encrypt()` method. 

The consequences is that the asymmetric implementation cannot make a difference between `SymmetricEncryptionInterface::encrypt($message, $key)` and `AsymmetricEncryptionInterface::encrypt($message, $public, $private = null)`. Ie, I need to rename the methods. 

This is all fine, it is currently a bit of a hack =)

Do you have a suggestion for new method names?

jderusse replied:
What's about `encrypt($message, Key $key)`
```php
SymmetricKey extends Key;
AsymmetricKey extends Key;
DoubleAsymmetricKey extends AsymmetricKey;
```

Nyholm replied:
Thank you. It took a day for this idea to sink in. 

I've updated the PR.

Nyholm commented:
I liked JWE because it mean that our payload could be portable. However, JWE introduced a lot of complexity and I didn't manage to add a Authentication Tag according to the spec. So.. just complexity and no probability. I changed back to a more simple implementation.

Nyholm commented:
Im trying out something new. Im making the API more explicit by introducing a KeyInterface. See 12804bb

sstok commented:
Sorry for this harsh comment, but it's noting personal: [DON'T EVER ROLL YOUR OWN ENCRYPTION!](https://security.stackexchange.com/questions/18197/why-shouldnt-we-roll-our-own)

I'm no crypto expert but I already see some things that are potentially dangerous.

- The message and private-key are passed-in as plain text, Halite requires that the them passed as a HiddenString to prevent stack traces from exposing sensitive information. 
- ` // Check if Authentication Tag is valid
        $hash = hash('sha256', $headersString.$payload.$nonce);`
   Authenticated encryption should use hmac-based hashing.
- `Ciphertext::create('RSAES-PKCS1-v1_5'` https://paragonie.com/blog/2018/04/protecting-rsa-based-protocols-against-adaptive-chosen-ciphertext-attacks (might be wrong on this one)

Why do you think Sodium was created? 😄 Cryptographic is a complex topic and with my limited knowledge I can already spot some things I'm not entirely sure of, but I know have caused major problems in the past. If you have sodium, use it. PHP has it bundled already, and it is the most secure standard of today. RSA and DSA should only used for legacy systems or by experts who know how the handle these complexities.

ping @paragonie-scott

jvasseur commented:
> Sorry for this harsh comment, but it's noting personal: [DON'T EVER ROLL YOUR OWN ENCRYPTION!](https://security.stackexchange.com/questions/18197/why-shouldnt-we-roll-our-own)

We are not creating our own encryption here but just creating a wrapper that relies on other lower level encryption libraries (including Sodium).

paragonie-security commented:
> We are not creating our own encryption here but just creating a wrapper that relies on other lower level encryption libraries (including Sodium).

We haven't looked at the code and offer no immediate opinion on its security or soundness, but I do want to offer something to think about in this discussion.

"Creating one's own encryption" isn't limited to designing your own primitives (e.g. block ciphers, stream ciphers, and hash functions). Any time you combine cryptographic primitives together, you're creating a cryptography protocol. It's tempting to think, "I'm only using secure components that I didn't roll myself, therefore I'm safe," but we have a name for that: [reasoning by lego](https://www.cryptofails.com/post/121201011592/bad-crypto-reasoning).

The layers you can possibly run into the "rolling your own crypto" criticism are:

* High-level protocols (TLS, Noise)
* Low-level protocols (ECIES, HPKE)
* Constructions (AES-GCM, AES-CTR + HMAC-SHA2)
* Primitives (AES, GMAC, SHA2)
* Theory (Mathematical constructions useful for pseudo-random functions, pseudo-random permutations, and/or asymmetric trapdoor functions with specific desired properties; e.g. supersingular isogenies; which can be used to design primitives).

Thus, saying, "we are not creating our own encryption here" is effectively, "we aren't designing our own primitives" or more charitably "we aren't designing our own constructions", but Symfony is *definitely* doing low-level cryptographic protocol design.

As Tay so eloquently puts it: "[Cryptography is nightmare magic math that cares what kind of pen you use.](https://twitter.com/SwiftOnSecurity/status/832058185049579524)"

Nyholm commented:
Thank you for the feedback and the links. 

I understand this PR is somewhat confusing. I started by adding a single class and then it grew to a new component without me updating the PR description. 

I've now updated the PR description so it will be clear what Im doing and why. I also think I show how all "security" is not handled by the component. Ie, If RSA is broken, it is not our problem but Phpseclib's problem. 

I would first ask for feedback on the abstraction (the interfaces). Are they good? Are they too strict or too flexible? Will they work?

After that feel free to comment on the implementations. I have followed their documentation and I am using Sodium and phpseclib as described.

sstok commented:
Ah. With the updated description It makes more sense 👍🏼 I was almost worried Symfony was rolling it's own crypto 😄

skoop commented:
From the use cases that I've so far used, the interfaces seem to be fine!

stof commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,137 @@
+     * kept a secret.
+     *
+     * When Alice and Bob wants to communicate they share their public keys with
+     * each other. Alice will encrypt a message with bobs public key. When Bob
```
typo: should be `Bob's`, not `bobs`

noniagriconomie commented on src/Symfony/Component/Encryption/KeyInterface.php:
```diff
@@ -0,0 +1,58 @@
+    public function toString(): string;
+
+    /**
+     * Creates a Key from stored data.
```
```suggestion
     * Creates a KeyInterface from stored data.
```

or 

```suggestion
     * Creates a key implementation from stored data.
```

stof commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,137 @@
+     *
+     * When Alice and Bob wants to communicate they share their public keys with
+     * each other. Alice will encrypt a message with bobs public key. When Bob
+     * receive the message, he will decrypt it with his private key.
```
`receives`

noniagriconomie commented on src/Symfony/Component/Encryption/Exception/DecryptionException.php:
```diff
@@ -0,0 +1,27 @@
+{
+    public function __construct($message = null, \Throwable $previous = null)
+    {
+        parent::__construct($message ?? 'Could not decrypt the ciphertext.', 0, $previous);
```
```suggestion
        parent::__construct($message ?? 'Could not decrypt the cipher text.', 0, $previous);
```

Nyholm replied:
It bugs be too, but it is "ciphertext". 
https://en.wikipedia.org/wiki/Ciphertext

noniagriconomie replied:
@Nyholm ha thanks, maybe then change inside Exception classes, you wrote `cipher text` :)

Nyholm replied:
oh =)

Sure.

stof commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,137 @@
+     *     $key = unserialize($bobPublicKey);
+     *     $ciphertext = $encryption->encryptFor('input', $key);
+     *
+     *     // Alice sends ciphertext to bob
```
`Bob`

stof commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,137 @@
+     *
+     * When Alice and Bob wants to communicate they share their public keys with
+     * each other. Alice will encrypt a message with keypair [ alice_private, bob_public ].
+     * When Bob receive the message, he will decrypt it with keypair [ bob_private, alice_public ].
```
`receives`

noniagriconomie commented on src/Symfony/Component/Encryption/Ciphertext.php:
```diff
@@ -0,0 +1,170 @@
+    private $nonce;
+
+    /**
+     * @var array additional headers
```
should you document that it is a key->value array?
related to the `getHeader()`

noniagriconomie commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,137 @@
+interface EncryptionInterface
+{
+    /**
+     * Generates new a Key to be used with encryption.
```
```suggestion
     * Generates new a KeyInterface to be used with encryption.
```

Nyholm replied:
I think this is correct. We are generating a "new Key" or a "new object implementing KeyInterface". But we are not generating "a new KeyInterface"

derrabus replied:
It should be `key` with a lower-case K then. 🙃

Nyholm replied:
Okey. 

Thank you

stof commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,137 @@
+     *     $keypair = $aliceKey->createKeypair($bobKey);
+     *
+     *     $ciphertext = $encryption->encryptForAndSign('input', $keypair);
+     *     $message = $encryption->decrypt($ciphertext, $bobKey->createKeypair($aliceKey));
```
this example looks broken to me. the decryption only uses the key of Alice, which is not what is described above

Nyholm replied:
Hm.. Maybe I should make the code more expressive. We do use Bob's key here: `$bobKey->createKeypair($aliceKey)`

noniagriconomie commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,137 @@
+     * @param string       $message plain text version of the message
+     * @param KeyInterface $myKey   a key that holds a string secret
+     *
+     * @return string the output
```
```suggestion
     * @return string the encrypted version of the message
```

as related to the other doc of `decrypt()`
same below i think

Nyholm replied:
I should name this format and document it. Then link to the documentation.

stof commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,137 @@
+     * <code>
+     *     $aliceKey = $encryption->generateKey();
+     *     $bobKey = $encryption->generateKey();
+     *     $keypair = $aliceKey->createKeypair($bobKey);
```
This looks wrong to me. When Alice generates the keypair for signing, using `$bobKey` should not be possible. Alice only knows Bob's public key, not bob's private key.

Nyholm replied:
It is correct. But I also see you point that it is confusing. 
I'll make it longer for clarity.

stof commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,137 @@
+     * </code>
+     *
+     * @param string       $message plain text version of the message
+     * @param KeyInterface $keypair A key with a public key of the recipient and a private key of the sender
```
Using `keypair` as name looks wrong to me. In all crypto articles I read, `keypair` refers to the pair of a private key and its associated public key.
Here, you use this for 2 unrelated keys (and you even mix them in the same object)

Nyholm replied:
Hmm. Im not sure if you are correct or not. 

What I want to refer to is a "pair of keys". It does not have to be a private keys corresponding public key. But it must be a private/public combination. 

Maybe find another name to make this more clear?

stof replied:
or maybe we should pass the recipient key and the sender key as separate arguments

Nyholm replied:
That would skip the "Key pair" issue.. Correct. Hm.. Yes that might be simpler

Good idea!

Nyholm replied:
But.. That also means that we need two keys for the `decrypt()`. 

Should I update it to : 

```php
EncryptionInterface::decrypt(string $message, KeyInterface $key, KeyInterface $senderPublicKey = null);
```
??

stof replied:
I'm wondering whether it is a good idea to have a single `decrypt` method covering the 3 encryption APIs. Implementations have to guess which of the 3 APIs was used to encrypt, because the decryption logic is different in each case.

Nyholm replied:
Both yes and no. 

The implementation will know what algorithm that has been used. Ie, it got a "routing key", so it does not have to guess.

noniagriconomie commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,137 @@
+    /**
+     * Generates new a Key to be used with encryption.
+     *
+     * Don't lose your private key and make sure to keep it a secret.
```
i would write first to make it secret (and where/how to "store" it)
then to not loosing it :) 

also and related to the secret component that has a key rotate feature, will this kind of api be present here?

Nyholm replied:
Yes, you can use this API to rotate the secrets. It is basically just to decrypt everything, generate a new key, encrypt everything, save the new key and throw away the old one.

stof commented on src/Symfony/Component/Encryption/Phpseclib/PhpseclibEncryption.php:
```diff
@@ -0,0 +1,172 @@
+use Symfony\Component\Encryption\KeyInterface;
+
+if (!class_exists(RSA::class)) {
+    throw new \LogicException('You cannot use "Symfony\Component\Security\Core\Encryption\PhpseclibEncryption" as the "phpseclib/phpseclib:2.x" package is not installed. Try running "composer require phpseclib/phpseclib:^2".');
```
this should be moved to the constructor instead. Otherwise, a `class_exists(PhpseclibEncryption::class)` could trigger it.

Nyholm replied:
Thank you

noniagriconomie commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,137 @@
+     *
+     * Asymmetric encryption uses a "key pair" ie a public key and a private key.
+     * It is safe to share your public key, but the private key should always be
+     * kept a secret.
```
```suggestion
     * kept secret.
```

noniagriconomie commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,137 @@
+     * It is safe to share your public key, but the private key should always be
+     * kept a secret.
+     *
+     * When Alice and Bob wants to communicate they share their public keys with
```
```suggestion
     * When Alice and Bob wants to communicate securely, they share their public keys with
```

noniagriconomie commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,137 @@
+     *     $bobKey = $encryption->generateKey();
+     *     $bobPublicKey = serialize($bobKey->createPublicKey());
+     *
+     *     // Bob sends the public key to Alice
```
```suggestion
     *     // Bob sends its public key to Alice
```

noniagriconomie commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,137 @@
+     * When Alice and Bob wants to communicate they share their public keys with
+     * each other. Alice will encrypt a message with bobs public key. When Bob
+     * receive the message, he will decrypt it with his private key.
+     *
```
as you mentioned the word "recipient" above, maybe make it clear here who is who in this process, sender, recipient
because they both "receive" things, either a key or a message

noniagriconomie commented on src/Symfony/Component/Encryption/Sodium/SodiumKey.php:
```diff
@@ -0,0 +1,178 @@
+use Symfony\Component\Encryption\Exception\InvalidKeyException;
+use Symfony\Component\Encryption\KeyInterface;
+
+final class SodiumKey implements KeyInterface
```
shouldnt those key implementations be internal?

Nyholm replied:
Yes, lets!

noniagriconomie commented on src/Symfony/Component/Encryption/LICENSE:
```diff
@@ -0,0 +1,19 @@
+Copyright (c) 2016-2020 Fabien Potencier
```
maybe remove 2016 no?

Nyholm replied:
Thank you

noniagriconomie commented on src/Symfony/Component/Encryption/KeyInterface.php:
```diff
@@ -0,0 +1,58 @@
+ * A Key for a specific user and specific Encryption implementation. Keys cannot
+ * be shared between Encryption implementations.
+ *
+ * A Key is always serializable.
```
why not implementing `Serializable`?

Nyholm replied:
Hm, yeah. That might be better.

noniagriconomie reviewed (commented):
reviews after the Slack post :)
will definitly use it

Nyholm commented:
Thank you for the reviews and comments. They are super helpful. 

Status: needs review

soatok commented on src/Symfony/Component/Encryption/Phpseclib/PhpseclibEncryption.php:
```diff
@@ -0,0 +1,175 @@
+            $aes->setPassword($myKey->getSecret());
+            $aes->setIV($nonce = Random::string($aes->getBlockLength() >> 3));
+
+            return Ciphertext::create('RSAES-PKCS1-v1_5', $aes->encrypt($message), $nonce)->getString();
```
Suggestion: Use OAEP instead of PKCS #1 v1.5.

https://framework.zend.com/security/advisory/ZF2015-10

Nyholm replied:
I've updated the name to better reflect what we actually control.

bendavies commented on composer.json:
```diff
@@ -135,7 +135,8 @@
         "twig/inky-extra": "^2.12",
-        "twig/markdown-extra": "^2.12"
+        "twig/markdown-extra": "^2.12",
+        "phpseclib/phpseclib": "^2.0.29"
```
these packages look mostly sorted, with one exception.

Nyholm replied:
Thank you

bendavies commented on src/Symfony/Component/Encryption/Phpseclib/PhpseclibKey.php:
```diff
@@ -0,0 +1,130 @@
+
+    public function createKeypair(KeyInterface $publicKey): KeyInterface
+    {
+        return self::fromPrivateAndPublicKeys($this->getPrivateKey(), $publicKey->getPublicKey());
```
`KeyInterface::getPublicKey()` does not exist

bendavies commented on src/Symfony/Component/Encryption/Sodium/SodiumKey.php:
```diff
@@ -0,0 +1,187 @@
+
+    public function createKeypair(KeyInterface $publicKey): KeyInterface
+    {
+        return self::fromPrivateAndPublicKeys($this->getPrivateKey(), $publicKey->getPublicKey());
```
`KeyInterface::getPublicKey()` does not exist

Nyholm replied:
Thank you.

bendavies commented on src/Symfony/Component/Encryption/Ciphertext.php:
```diff
@@ -0,0 +1,170 @@
+    public static function parse(string $input): self
+    {
+        $parts = explode('.', $input);
+        if (false === $parts || 4 !== \count($parts)) {
```
false will never be returned here?
`If separator is an empty string (""), explode() will return false.`

Nyholm replied:
THank you

bendavies commented on src/Symfony/Component/Encryption/Phpseclib/PhpseclibEncryption.php:
```diff
@@ -0,0 +1,175 @@
+            // Load private key after encryption
+            $rsa->loadKey($keypair->getPublicKey());
+            $rsa->setSignatureMode(RSA::SIGNATURE_PSS);
+            $headers['signature'] = base64_encode($rsa->sign($ciphertext));
```
I've never seen this style?
```php
$headers = ['signature' => base64_encode($rsa->sign($ciphertext))];
```

Nyholm replied:
What do you mean?

bendavies replied:
I mean, i've never seen an array defined the way you have here

Nyholm replied:
```php
$headers = [];
$headers = ['foo'=>'bar'];
$headers = ['foo' => json_encode('bar')];
```

Sure I could write something like this, but that is an extra line. 

```php 
$headers = [];
$headers['signature'] = base64_encode($rsa->sign($ciphertext));
```

stof commented on src/Symfony/Component/Encryption/KeyInterface.php:
```diff
@@ -0,0 +1,48 @@
+ *
+ * @experimental in 5.3
+ */
+interface KeyInterface extends \Serializable
```
please don't enforce the usage of `\Serializable`. It is not the only way to make a PHP serializable (and it is the worse one actually)

Nyholm replied:
Could you help me undertand what I should use instead?

I dont want magic methods in the interface, do I?

stof replied:
I would keep it only in the phpdoc, and let each implementation decide what they want to use

Nyholm replied:
Done, thank you

jderusse commented on src/Symfony/Component/Encryption/.gitattributes:
```diff
@@ -0,0 +1,3 @@
+/.gitignore export-ignore
+/phpunit.xml.dist export-ignore
```
missing `.gitattributes` (see other components)

jderusse commented on src/Symfony/Component/Encryption/Ciphertext.php:
```diff
@@ -0,0 +1,170 @@
+            throw new MalformedCipherException();
+        }
+
+        $headers = json_decode($headersString, true);
```
what if `headersString` is an invalid json?

Nyholm replied:
It shouldn't be (we verified the string with the hash). But If we created an invalid json string then did the hash, the $headers non me an array, we have a check for that on the next line.

jderusse commented on src/Symfony/Component/Encryption/Ciphertext.php:
```diff
@@ -0,0 +1,170 @@
+        }
+
+        $headers = json_decode($headersString, true);
+        if (!\is_array($headers) || !\array_key_exists('alg', $headers)) {
```
You should also assert `ver === 1`

Nyholm replied:
Thank you

jderusse commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,156 @@
+     * @throws EncryptionException
+     * @throws InvalidKeyException
+     */
+    public function encrypt(string $message, KeyInterface $myKey): string;
```
hmm `$key` no? (especially with php8 + namedparameters)

Nyholm replied:
Yes, sure.

jderusse commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,156 @@
+     * @throws EncryptionException
+     * @throws InvalidKeyException
+     */
+    public function encryptFor(string $message, KeyInterface $recipientKey): string;
```
symmetric or asymmetric encryption could be guessed from the type of key right? We could use the same method

Nyholm replied:
Actually no. 

A `KeyInterface` will internally always create a symmetric key and an asymmetric key. But that knowledge is hidden inside the implementation. 

I also like the fact with explicit names atm (Yes, I know that I only had one `encrypt()` method 3 days ago).

jderusse commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,156 @@
+     * @throws EncryptionException
+     * @throws InvalidKeyException
+     */
+    public function encryptForAndSign(string $message, KeyInterface $recipientKey, KeyInterface $senderKey): string;
```
I'd rather `$signer->sign($encrypter->encrypt($message, $recipientPublicKey), $myPrivateKey)`
It' fine the `sign` plain text, see how JWT works

Nyholm replied:
If we have a dedicated signer interface, then the sodium implementation requires you to have 2 keys. One for encryption and one for signing. 

That is why I dont really like to separate the interface. I also think it brings complexity that is not really needed. I mean, yes, you can sign a plain text just like JWT is doing. But what is the benefit? Is it to reduce the cost of encryption?

chalasr replied:
I'm fine with this method as an high-level API, the one we would suggest the most in the docs, though I agree it seems to be worth splitting encryption and signature at a lower level. 
Providing both APIs would be the best I think. Probably not a blocker for this PR though

Nyholm replied:
If we are going to use Sodium AND if we are going to have one api for encryption and one for signing, then we will need two sets of keys. Eg, "encryption key" and "signing key". 

Im currently on the fence for this with the argument that this would be an easy to use abstraction. If you need to "only sign", you are the "advanced use case" and you will have to use sodium directly.

chalasr replied:
Using the same key pair for both encryption and signature is generally a bad idea, as the requirements for signature keys are often different that encryption ones in terms of e.g. backups, rotation or revokation. 
I suggest to reconsider :)

Nyholm replied:
Hm. But now you open a new problem of key management that I conveniently excluded from this PR. 

Should we still leave this to the user or should we help them somehow?

chalasr replied:
Sorry but I think it's more than a key management issue. 
From https://libsodium.gitbook.io/doc/quickstart#how-can-i-sign-and-encrypt-using-the-same-key-pair:

> If you really need to sign and encrypt a message, possibly for signatures to be publicly verified, consider using [signcryption](https://github.com/jedisct1/libsodium-signcryption).
[...]
However, this is not recommended and usually not necessary. Prefer using distinct key pairs instead.

Considering this and:

>  In most cases, signing a message in addition to encrypting it is not required.

 I'd say we shouldn't sign at all.

jderusse commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,156 @@
+     * @throws DecryptionException
+     * @throws InvalidKeyException
+     */
+    public function decrypt(string $message, KeyInterface $key, KeyInterface $senderPublicKey = null): string;
```
`$encrypter->decrypt($signer->verify($message, $myPrivateKey), $sendPublicKey)`

jderusse commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/encryption.php:
```diff
@@ -0,0 +1,35 @@
+
+        ->set('security.encryption.sodium', SodiumEncryption::class)
+            ->args([
+                '%kernel.secret%',
```
not needed anymore right? (same below)

Nyholm replied:
Correct. Thank you

jderusse commented on src/Symfony/Component/Encryption/Phpseclib/PhpseclibEncryption.php:
```diff
@@ -0,0 +1,185 @@
+            // Load private key after encryption
+            $rsa->loadKey($senderKey->getPrivateKey());
+            $rsa->setSignatureMode(RSA::SIGNATURE_PSS);
+            $headers['signature'] = base64_encode($rsa->sign($ciphertext));
```
Why do we `base64_encode`? This should be the job of `Ciphertext`?

Nyholm replied:
json_encode cannot encode non ascii strings. 

So I could loop over `$headers` and run `base64_encode` on all values before I json_encode it... but Im not sure that is great. This seams eaasier.

jderusse commented on src/Symfony/Component/Encryption/Phpseclib/PhpseclibEncryption.php:
```diff
@@ -0,0 +1,185 @@
+        set_error_handler(__CLASS__.'::throwError');
+
+        try {
+            $aes = new AES(AES::MODE_CBC);
```
Is it ok? I'm not expert in cryptography.
same for all other choices in this class. (iv length, RSA encryptionMode, SignatureMode)

Nyholm replied:
Sort answer: I think so, but Im not sure. 

This is basically copied from their documentation.

jderusse commented on src/Symfony/Component/Encryption/Sodium/SodiumKey.php:
```diff
@@ -0,0 +1,176 @@
+
+        $secretLength = \strlen($secret);
+        if ($secretLength > \SODIUM_CRYPTO_SECRETBOX_KEYBYTES) {
+            $key->secret = substr($secret, 0, \SODIUM_CRYPTO_SECRETBOX_KEYBYTES);
```
is it safe to cut the user's input like that?

Nyholm replied:
It is, but we dont have any other option. Sodium requires the key length to be exactly SODIUM_CRYPTO_SECRETBOX_KEYBYTES.

Nyholm replied:
Hm.. But if you have a 40 char long key, and then just add a char at the end to invalidate it, it will not work.. 
I'll update this section.

jderusse commented on src/Symfony/Component/Encryption/Phpseclib/PhpseclibEncryption.php:
```diff
@@ -0,0 +1,185 @@
+            $rsa->setSignatureMode(RSA::SIGNATURE_PSS);
+            $headers['signature'] = base64_encode($rsa->sign($ciphertext));
+
+            return Ciphertext::create('RSA-OAEP-PSS', $ciphertext, random_bytes(8), $headers)->getString();
```
Ciphertext exposes an almost not used headers parameter and a nonce that is not always necessary.

I've the feeling that something is wrong in this class.

Nyholm replied:
The Chipertext is very generic. It supports the new fancy encryption algorithm of tomorrow.

Any suggestions for improvement is welcomed.

jderusse commented on src/Symfony/Component/Encryption/Ciphertext.php:
```diff
@@ -0,0 +1,170 @@
+            self::base64UrlEncode($headers),
+            self::base64UrlEncode($this->payload),
+            self::base64UrlEncode($this->nonce),
+            self::base64UrlEncode(hash('sha256', $headers.$this->payload.$this->nonce))
```
What is the purpose of this signature?

Nyholm replied:
It is an integrity check to make sure we have received the full string. 

This is less useful when you send a message over TCP but is more useful when you store the key on disk or in db. Basically a quick sanity check.

nicolas-grekas replied:
I'm also wondering about this signature. Doesn't the encrypted payload already handle its own integrity?

Nyholm replied:
Yes, the payload does. But not the header and nonce.

jderusse replied:
hmm. in that case, the same nonce is used twice. Isn't it a weakness?

do we really need this headers property? I feel the code and the api would be more simpler without it. 
people can json_encode their own array if they want providing extra data

jderusse reviewed (commented):
Few comments.

I'm still not convince by providing a `Encryption` component.

Nyholm commented:
Thank you for the reviews. I've updated the PR accordingly

Status: Needs review

OskarStark commented on src/Symfony/Component/Encryption/Exception/DecryptionException.php:
```diff
@@ -0,0 +1,27 @@
+ */
+class DecryptionException extends \RuntimeException implements ExceptionInterface
+{
+    public function __construct($message = null, \Throwable $previous = null)
```
```suggestion
    public function __construct(string $message = null, \Throwable $previous = null)
```

OskarStark commented on src/Symfony/Component/Encryption/Exception/EncryptionException.php:
```diff
@@ -0,0 +1,27 @@
+ */
+class EncryptionException extends \RuntimeException implements ExceptionInterface
+{
+    public function __construct($message = null, \Throwable $previous = null)
```
```suggestion
    public function __construct(string $message = null, \Throwable $previous = null)
```

OskarStark commented on src/Symfony/Component/Encryption/Exception/InvalidKeyException.php:
```diff
@@ -0,0 +1,27 @@
+ */
+class InvalidKeyException extends \RuntimeException implements ExceptionInterface
+{
+    public function __construct($message = null, \Throwable $previous = null)
```
```suggestion
    public function __construct(string $message = null, \Throwable $previous = null)
```

OskarStark commented on src/Symfony/Component/Encryption/Exception/MalformedCipherException.php:
```diff
@@ -0,0 +1,25 @@
+ */
+class MalformedCipherException extends DecryptionException
+{
+    public function __construct($message = null, \Throwable $previous = null)
```
```suggestion
    public function __construct(string $message = null, \Throwable $previous = null)
```

OskarStark commented on src/Symfony/Component/Encryption/LICENSE:
```diff
@@ -0,0 +1,19 @@
+Copyright (c) 2020-2020 Fabien Potencier
```
```suggestion
Copyright (c) 2020 Fabien Potencier
```

OskarStark commented on src/Symfony/Component/Encryption/Tests/AbstractEncryptionTest.php:
```diff
@@ -0,0 +1,174 @@
+        $ciphertext = $cipher->encrypt($message, $key);
+
+        $key2 = $cipher->generateKey('s3cr3t');
+        $this->assertEquals($message, $cipher->decrypt($ciphertext, $key2));
```
Not sure, but can we use `assertSame()` here?

OskarStark commented on src/Symfony/Component/Encryption/Tests/Phpseclib/PhpseclibEncryptionTest.php:
```diff
@@ -0,0 +1,39 @@
+/**
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+class PhpseclibEncryptionTest extends AbstractEncryptionTest
```
I would make the tests final, but just a personal preference

nicolas-grekas replied:
we don't really care about putting `final` on tests: that does have any impact on maintenance

OskarStark commented on src/Symfony/Component/Encryption/Ciphertext.php:
```diff
@@ -0,0 +1,184 @@
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ *
+ * @experimental in 5.3
+ *
+ * @internal
```
It is only experimental if it is public, I think it is not needed to mark it as experimental if it is internal and vice versa 🤔

OskarStark commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,156 @@
+     * </code>
+     *
+     * @param string       $message      plain text version of the message
+     * @param KeyInterface $recipientKey A public key of the recipient
```
```suggestion
     * @param KeyInterface $recipientKey public key of the recipient
```

OskarStark commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,156 @@
+     *
+     * @param string       $message      plain text version of the message
+     * @param KeyInterface $recipientKey A public key of the recipient
+     * @param KeyInterface $senderKey    A private key of the sender
```
```suggestion
     * @param KeyInterface $senderKey    private key of the sender
```

OskarStark commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,156 @@
+     * Get a plain text version of the encrypted message.
+     *
+     * @param string       $message         encrypted version of the message
+     * @param KeyInterface $key             The key of the recipient. It should contain a private key
```
```suggestion
     * @param KeyInterface $key             key of the recipient, it should contain a private key
```

OskarStark commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,156 @@
+     * </code>
+     *
+     * @param string       $message      plain text version of the message
+     * @param KeyInterface $recipientKey A key with a public key of the recipient
```
```suggestion
     * @param KeyInterface $recipientKey key with a public key of the recipient
```

OskarStark commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,156 @@
+     * Get an encrypted version of the message that only the recipient can read.
+     * The recipient can also verify who sent the message.
+     *
+     * Asymmetric encryption uses a "key pair" ie a public key and a private key.
```
```suggestion
     * Asymmetric encryption uses a "key pair" i.e. a public key and a private key.
```

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/encryption.php:
```diff
@@ -0,0 +1,24 @@
+    $container->services()
+        ->set('security.encryption.sodium', SodiumEncryption::class)
+        ->set('security.encryption.phpseclib', PhpseclibEncryption::class)
+        ->alias(EncryptionInterface::class, 'security.encryption.phpseclib')
```
shouldn't this use sodium by default when available?
since sodium is now bundled by default, that'd make more sense to me.
we should suggest somewhere to install either sodium or phpseclib when enabling the component

Nyholm replied:
We definitely should, I must have made a misstake when refactoring this.

nicolas-grekas commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,156 @@
+     *
+     * Don't lose your private key and make sure to keep it a secret.
+     *
+     * @param string|null $secret to be used in symmetric encryption. A new secret
```
description of arguments should all start with an uppercase letter (same in all docblocks)

nicolas-grekas commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,156 @@
+    public function generateKey(string $secret = null): KeyInterface;
+
+    /**
+     * Get an encrypted version of the message.
```
Missing "s": Gets (same in all docblocks)

nicolas-grekas commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,156 @@
+     * @param KeyInterface $recipientKey public key of the recipient
+     * @param KeyInterface $senderKey    private key of the sender
+     *
+     * @return string the output
```
can't we remove the `@param` and `@return`? (we should wonder the same in all docblocks)
maybe rename the args to `$recipientPublicKey` / `$senderPrivateKey` since that "private"/"public" might be the useful part of the description?

Nyholm replied:
> can't we remove the @param and @return?

I want to add some name of the output, previously it said "the JWE output" . 

> maybe rename the args to $recipientPublicKey / $senderPrivateKey

Hm.. For these implementations, you are correct. But I dont have the concept of "public" and "private" key in the current PR. We only have "key" and "public key".

nicolas-grekas commented on src/Symfony/Component/Encryption/Sodium/SodiumKey.php:
```diff
@@ -0,0 +1,173 @@
+use Symfony\Component\Encryption\KeyInterface;
+
+/**
+ * @internal
```
we could do like we do in the Cache component: use a rebound closure in `SodiumEncryption::generateKey()` to forbid creating this class in userland.

nicolas-grekas commented on src/Symfony/Component/Encryption/Sodium/SodiumKey.php:
```diff
@@ -0,0 +1,173 @@
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+final class SodiumKey implements KeyInterface, \Serializable
```
Serializable is a dead interface, it should never ever be used anymore
See also https://wiki.php.net/rfc/phase_out_serializable

nicolas-grekas commented on src/Symfony/Component/Encryption/Tests/AbstractEncryptionTest.php:
```diff
@@ -0,0 +1,174 @@
+
+        $ciphertext = $cipher->encryptForAndSign($input = 'input', $bobKey->extractPublicKey(), $aliceKey);
+        $this->expectException(SignatureVerificationRequiredException::class);
+        $this->assertEquals($input, $cipher->decrypt($ciphertext, $this->createPrivateKey($bobKey)));
```
let's use assertSame instead of assertEquals as much as possible

nicolas-grekas reviewed (commented):
Why do we need a phpseclib driver?
Can't we rely on sodium_compat for ppl that don't have sodium enabled?
Here are more random notes :)

Nyholm commented:
> Why do we need a phpseclib driver?

That is a good observation. There are only two minor reasons to have phpseclib; 1) It helps me see if it is a good abstraction by providing to different implementations, 2) If you are communicating with non-php app, then RSA is way more nicer that sodium. 

I'll make sure to update the PR with suggestions from Oskar and Nicolas. 

Status: needs work

Nyholm commented:
Thank you for the review. I've updated the PR. 

Status: needs review

derrabus commented on composer.json:
```diff
@@ -126,6 +126,7 @@
         "paragonie/sodium_compat": "^1.8",
         "pda/pheanstalk": "^4.0",
         "php-http/httplug": "^1.0|^2.0",
+        "phpseclib/phpseclib": "^2.0.29",
```
There's a version 3 already.

derrabus commented:
I think having this component is a good idea, if it allows us to replace encryption-related logic from other components. 👍🏻

terrafrost commented:
> Why do we need a phpseclib driver?
> Can't we rely on sodium_compat for ppl that don't have sodium enabled?
> Here are more random notes :)

https://phpseclib.com/docs/why#phpseclib-vs-libsodium addresses this

jderusse commented on src/Symfony/Component/Encryption/Sodium/SodiumKey.php:
```diff
@@ -0,0 +1,163 @@
+    public function getPrivateKey(): string
+    {
+        if (null === $this->privateKey) {
+            throw new InvalidKeyException('This key does not have a private key.');
```
you can fallback on keypair

jderusse commented on src/Symfony/Component/Encryption/Sodium/SodiumKey.php:
```diff
@@ -0,0 +1,163 @@
+    public static function fromKeypair(string $keypair): self
+    {
+        $key = new self();
+        $key->keypair = $keypair;
```
what's about publicKey and privateKey (they are initialized in create method but not here)

Nyholm replied:
Yeah. Sure. 

Thank you. This will solve your other comments.

jderusse commented on src/Symfony/Component/Encryption/Sodium/SodiumKey.php:
```diff
@@ -0,0 +1,163 @@
+    public function getPublicKey(): string
+    {
+        if (null === $this->publicKey) {
+            throw new InvalidKeyException('This key does not have a public key.');
```
you can fallback on privateKey or keyPair

Nyholm commented:
I added a super small demo application to show how one would encrypt properties on an entity. Ie to store an encrypted value for `User::$email` in the database. 

https://github.com/Nyholm/symfony-encrypter-demo

fabpot commented:
@Nyholm What's the status here? Are you willing to finish the first version of this new component?

OskarStark commented:
If for 6.0 you should adjust the PR header and the milestone

Nyholm commented:
Thank you. 

I want it to be experimental. We cannot have experimental code in 5.4. 

I've updated the milestone, what do you mean with "PR header"?

EDIT: 
You mean the PR description. Got it!

chalasr commented:
IMHO we should save us from maintaining a phpseclib implementation. 
Providing one built-in implementation is enough to start and sodium is widely supported out there, while the flexibility of RSA  is calling for trouble.

Nyholm commented:
I removed the phpseclib implementation. 

Psalm yields a false negative. I also think the PHPUnit 8.0 high-deps is a false negative. The CI config does not like it when I added a new component. 

Status: Needs review.

Nyholm commented on src/Symfony/Component/Encryption/Ciphertext.php:
```diff
@@ -0,0 +1,183 @@
+        $model = new self();
+        $model->algorithm = $algorithm;
+        $model->payload = $ciphertext;
+        $model->nonce = $nonce;
+        $model->headers = $headers;
```
The Nonce is used for most (not all) encryption. We could move it into the headers instead. That would mean that `headers` amore more like `context`.

chalasr replied:
sounds good to me

Nyholm replied:
I started to remove the nonce, but I change my mind. 

The headers must be ascii characters because we are json encoding it. 
However, good nonces do not restrict themselves to ascii.

chalasr commented on src/Symfony/Component/Encryption/Ciphertext.php:
```diff
@@ -0,0 +1,183 @@
+    }
+
+    /**
+     * Take a string representation of the chiphertext and parse it into an object.
```
```suggestion
     * Take a string representation of the ciphertext and parse it into an object.
```

chalasr commented on src/Symfony/Component/Encryption/Exception/UnsupportedAlgorithmException.php:
```diff
@@ -0,0 +1,25 @@
+{
+    public function __construct(string $algorithm, \Throwable $previous = null)
+    {
+        parent::__construct(sprintf('The ciphertext is encrypted with "%s" algorithm. Decryption of that algorithm is not supported.', $algorithm));
```
Providing the list of supported algorithms would be useful if possible

chalasr commented on src/Symfony/Component/Encryption/Exception/SignatureVerificationRequiredException.php:
```diff
@@ -0,0 +1,28 @@
+{
+    public function __construct(\Throwable $previous = null)
+    {
+        parent::__construct('The sender requires you to verify the signature.');
```
just looking at the message, it's not obvious to me what needs to be done to get rid of this error. Could we suggest passing the missing sender public key when it's thrown from `decrypt()`?

chalasr commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,155 @@
+interface EncryptionInterface
+{
+    /**
+     * Generates new a key to be used with encryption.
```
```suggestion
     * Generates a new key to be used for encryption.
```

chalasr commented on src/Symfony/Component/Encryption/Ciphertext.php:
```diff
@@ -0,0 +1,183 @@
+ *
+ * @internal
+ */
+class Ciphertext
```
```suggestion
class Ciphertext implements \Stringable
```

chalasr commented on src/Symfony/Component/Encryption/Ciphertext.php:
```diff
@@ -0,0 +1,183 @@
+        return $model;
+    }
+
+    /**
+     * @return string
+     */
+    public function __toString()
```
```suggestion
    public function __toString(): string
```

chalasr commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,155 @@
+    /**
+     * Gets an encrypted version of the message that only the recipient can read.
+     *
+     * Asymmetric encryption uses a "key pair" ie a public key and a private key.
```
```suggestion
     * Asymmetric encryption uses a "key pair" i.e. a public key and a private key.
```

chalasr commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,155 @@
+     * Gets an encrypted version of the message that only the recipient can read.
+     *
+     * Asymmetric encryption uses a "key pair" ie a public key and a private key.
+     * It is safe to share your public key, but the private key should always be
```
```suggestion
     * It is safe to share the public key, but the private key should always be
```

chalasr commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,155 @@
+     * It is safe to share your public key, but the private key should always be
+     * kept a secret.
+     *
+     * When Alice and Bob wants to communicate securely, they share their public keys with
```
```suggestion
     * When Alice and Bob want to communicate securely, they share their public keys with
```

chalasr commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,155 @@
+     * @param string       $message Plain text version of the message
+     * @param KeyInterface $key     A key that holds a string secret
+     *
+     * @return string Output formatted by Ciphertext
```
Given `Ciphertext` is an internal implementation detail (stop me if I'm wrong), I think the contract should not mention it. I'd replace this something briefly explaining what kind of string should be returned

Nyholm replied:
I agree, it make sense. I'll update

Nyholm replied:
You are correct that `Ciphertext` class should not really be mentioned. But the format of the string falls under or BC promise. So we need to call that string something... ie "Symfony Encryption Token". 

We also need to define (in document, not just code) what a SET is. Should I rename the class `Ciphertext` to `SymfonyEncryptionToken`? After that we can refer to "Symfony Encryption Token" (the format, not the class).

chalasr replied:
Do we really need to specify it more than just `string` after all? Looking at other crypto libs like paragonie/halite or defuse/encryption, no one seems to enforce the format of the encrypted message

Nyholm replied:
If we want the final string to be transferable to other non-symfony applications, then yes. 

The format of our string will fall under or BC promise.

chalasr commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,155 @@
+     * @param KeyInterface $recipientKey Public key of the recipient
+     * @param KeyInterface $senderKey    Private key of the sender
+     *
+     * @return string Output formatted by Ciphertext
```
same here

chalasr commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,155 @@
+     * It is safe to share your public key, but the private key should always be
+     * kept secret.
+     *
+     * When Alice and Bob wants to communicate securely, they share their public keys with
```
```suggestion
     * When Alice and Bob want to communicate securely, they share their public keys with
```

chalasr commented on src/Symfony/Component/Encryption/Exception/SignatureVerificationRequiredException.php:
```diff
@@ -0,0 +1,28 @@
+
+/**
+ * The sender requires you to verify the signature. You should pass both your
+ * private key and the senders public key to the decrypt() method.
```
```suggestion
 * private key and the sender's public key to the decrypt() method.
```

chalasr reviewed (changes requested):
Really nice work.

Nyholm committed:
Apply suggestions from code review

Co-authored-by: Robin Chalas <chalasr@users.noreply.github.com>

noniagriconomie commented on src/Symfony/Component/Encryption/Ciphertext.php:
```diff
@@ -0,0 +1,180 @@
+    /**
+     * @var string algorithm for the encryption
+     */
+    private $algorithm;
```
```suggestion
    private string $algorithm;
```

wdyt?
I've seen some PRs targeting v6 adding this kind of code for existing codebase,
thus maybe adding it directly to new one is a good option :)

Nyholm replied:
Good idea. I've update the PR

Nyholm committed:
Fixed exceptions

Nyholm committed:
Use PHP8 syntax

Nyholm committed:
Minor fixes

derrabus commented on src/Symfony/Component/Encryption/Sodium/SodiumKey.php:
```diff
@@ -0,0 +1,152 @@
+
+    public function getSecret(): string
+    {
+        if (null === $this->secret) {
+            throw new InvalidKeyException('This key does not have a secret.');
+        }
+
+        return $this->secret;
```
```suggestion
        return $this->secret ?? throw new InvalidKeyException('This key does not have a secret.');
```

PHP 8 🎉

Nyholm replied:
Hm.. Im not a big fan of this to be honest... Maybe because it is new...

¯\\_(ツ)_/¯

derrabus replied:
I really like that feature for situations like this. There are two more getters that can be transformed similarly.

derrabus commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/encryption.php:
```diff
@@ -0,0 +1,22 @@
+
+return static function (ContainerConfigurator $container) {
+    $container->services()
+        ->set('security.encryption.sodium', SodiumEncryption::class)
```
```suggestion
        ->set('encryption.sodium', SodiumEncryption::class)
```

I'd skip the `security` prefix because this is not a security sub-component.

Shall we add an alias `encryption` as well, to allow explicit wiring of the service?

Nyholm replied:
Are we doing aliases like `encryption`? 

Isn't the alias for `EncryptionInterface::class` enough?

------

I've removed `security.`

derrabus replied:
We usually do, if I'm not mistaken. I find these aliases still helpful for scenarios where I don't use autowiring.

Nyholm committed:
Apply suggestions from code review

Co-authored-by: Alexander M. Turek <me@derrabus.de>

Nyholm committed:
minior

soatok commented on src/Symfony/Component/Encryption/Ciphertext.php:
```diff
@@ -0,0 +1,172 @@
+        $nonce = self::base64UrlDecode($nonce);
+        $hashSignature = self::base64UrlDecode($hashSignature);
+
+        // Check if integrity hash is valid
```
This doesn't provide integrity. HMAC-SHA256 with a secret key would provide integrity. This just tells you if an error occurred.

soatok commented on src/Symfony/Component/Encryption/Sodium/SodiumEncryption.php:
```diff
@@ -0,0 +1,125 @@
+ */
+final class SodiumEncryption implements EncryptionInterface
+{
+    public function generateKey(string $secret = null): KeyInterface
```
`$secret` is not a good name for this parameter. Maybe `$symmetric`?

Nyholm replied:
Hm.. Yes. A user may wrongly assume that if they generate two keys with the same secret, the keys would be identical. 

I dont like `$symmetric` because it is an adjective and it also leaks implementation details. 

Btw, the `SodiumKey` has a method `fromSecret()`.

soatok replied:
Symmetric isn't a leaked implementation detail. You fundamentally have to treat Asymmetric Encryption and Symmetric Encryption as separate things when you deal with encryption.

See https://paragonie.com/blog/2015/08/you-wouldnt-base64-a-password-cryptography-decoded for more.

Nyholm replied:
This is not encryption. This is an abstraction over encryption. 

As you can see, the implementation is very strict with separating symmetric and asymmetric encryption.

soatok replied:
Even still, `$secret` is a bad parameter name here.

soatok replied:
Let me elaborate:

If you're generating an asymmetric key pair, it consists of a secret key and a public key.
If you're generating a symmetric key, you only have that one key. 

Which means: That you're **still always generating a secret key**, regardless of that boolean's state.

"Symmetric" or "shared" are better names for the parameter, which also needs to percolate down to the interface and any classes that implement it.

noniagriconomie commented on src/Symfony/Component/Encryption/README.md:
```diff
@@ -0,0 +1,20 @@
+====================
+
+The Encryption Component is an opinionated, high level abstraction over encryption
+libraries and PHP's Sodium extension. Encryption should be understandable without
```
`over encryption libraries and PHP's Sodium extension`
I see the sodium but not `encryption libraries`, maybe a leftover a new dev?

Nyholm committed:
From suggestion by Alexander

Nyholm committed:
updated comment

Nyholm committed:
Readme update

chalasr commented on src/Symfony/Component/Encryption/README.md:
```diff
@@ -0,0 +1,19 @@
+====================
+
+The Encryption Component is an opinionated, high level abstraction over PHP's Sodium
+extension. Encryption should be understandable without an degree in Cryptography.
```
```suggestion
extension. Encryption should be understandable without a degree in Cryptography.
```

Nyholm committed:
Remove signing

Nyholm committed:
Move nonce to the cyphertext headers

Nyholm committed:
Rename cihpertext

Nyholm commented:
Thank you all for the feedback. 
I've updated the PR accordingly. I think I've answered or made a change for all discussions. 

Status: needs review

soatok commented:
If your goal is to have a minimalistic encryption interface, what you've implemented here certainly meets that goal. However, you will find (as evidenced by the remarks about key management) that a lot of enterprise use-cases for encryption require a layered construction generally called ["envelope encryption"](https://docs.aws.amazon.com/kms/latest/developerguide/concepts.html#enveloping).

It's very easy to build an envelope encryption atop what you've already built, so this can be "future scope", but I wanted to bring it up now rather than have this surprise you later.

chalasr commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/encryption.php:
```diff
@@ -0,0 +1,23 @@
+        ->set('encryption.sodium', SodiumEncryption::class)
+        ->alias(EncryptionInterface::class, 'encryption.sodium')
+        ->alias('encryption', 'encryption.sodium')
+        ;
```
```suggestion
    ;
```

Nyholm replied:
Thank you

chalasr reviewed (approved):
This looks good to me.

Nyholm committed:
Update src/Symfony/Bundle/FrameworkBundle/Resources/config/encryption.php

Co-authored-by: Robin Chalas <chalasr@users.noreply.github.com>

jderusse commented on src/Symfony/Component/Encryption/EncryptionInterface.php:
```diff
@@ -0,0 +1,110 @@
+     * @throws EncryptionException
+     * @throws InvalidKeyException
+     */
+    public function encryptFor(string $message, KeyInterface $recipientKey): string;
```
do we really need 2 methods. The doc bloc says symmetric encryption is weaker.
Shouldn't we remove it ?

Nyholm replied:
Symmetric encryption is theoretically weaker but it still make sense to keep around, doesn't it? 

Using the full framework the key will default to `APP_SECRET`. It will make it easy when you just want to "hide values". Ie, dont show email in the URL when sending login links etc.

jderusse commented on src/Symfony/Component/Encryption/Sodium/SodiumKey.php:
```diff
@@ -0,0 +1,140 @@
+        return $key;
+    }
+
+    public static function fromPrivateAndPublicKeys(string $privateKey, string $publicKey): self
```
not used and class is internal

Nyholm replied:
I believe these method will be very useful when we use the encryption component for the "symfony secrets".

chalasr replied:
I have the feeling that having some usages of this new component in core would help moving forward here. 
@Nyholm Maybe could you open a PR doing so based on this PR? Or commit here directly?

jderusse commented on src/Symfony/Component/Encryption/Sodium/SodiumKey.php:
```diff
@@ -0,0 +1,140 @@
+        return $key;
+    }
+
+    public static function fromKeypair(string $keypair): self
```
same

jderusse commented on src/Symfony/Component/Encryption/SymfonyEncryptionToken.php:
```diff
@@ -0,0 +1,165 @@
+        $headers['alg'] = $this->algorithm;
+        $headers['ver'] = (isset($headers['ver']) && '' !== $headers['ver']) ? $headers['ver'] : '1';
+        foreach ($headers as $name => $value) {
+            $headers[$name] = self::base64UrlEncode($value);
```
`headers` is json_encoded + bade64encoded, no need to base64 encode the value twice IMHO

Nyholm replied:
Actually, we do. 

`json_encode` does not support non-ascii values. We do `base64UrlEncode()` on each value to make sure we only have ascii values so we are able to run json_encode. 

It is common to have nonce with non-ascii values.

paragonie-security commented on src/Symfony/Component/Encryption/Sodium/SodiumKey.php:
```diff
@@ -0,0 +1,140 @@
+            $key->secret = $secret;
+        } else {
+            // Trim the key to a good size
+            $key->secret = substr(sha1($secret), 0, \SODIUM_CRYPTO_SECRETBOX_KEYBYTES);
```
First, [`sha1()` is not an appropriate function to use here](https://shattered.io).

But even worse, this is hex-encoded so you're getting roughly *half* of the expected entropy out of your secret from this sha1 invocation.

```diff
- $key->secret = substr(sha1($secret), 0, \SODIUM_CRYPTO_SECRETBOX_KEYBYTES);
+ $key->secret = mb_substr(hash('sha256', $secret, true), 0, \SODIUM_CRYPTO_SECRETBOX_KEYBYTES, '8bit');
```

paragonie-security commented on src/Symfony/Component/Encryption/SymfonyEncryptionToken.php:
```diff
@@ -0,0 +1,165 @@
+
+        // Check if data has been modified
+        $hash = hash('sha256', $headersString.$ciphertext);
+        if (!hash_equals($hash, $hashSignature)) {
```
Misnomer: `$hashSignature` isn't, [at all](https://paragonie.com/blog/2015/08/you-wouldnt-base64-a-password-cryptography-decoded), a hash signature.

If you were doing the right thing, it would be a MAC, not a signature, but this just a SHA256 hash *without* any secrets. This is trivially bypassed: Just calculate `hash('sha256', $headersString.$ciphertext)` for any arbitrary data you want to feed in.

You're probably better off calling this a checksum, or equivalent, so as to not over-sell the security properties you're shooting for.

soatok replied:
This setup also looks susceptible to [canonicalization attacks](https://soatok.blog/2021/07/30/canonicalization-attacks-against-macs-and-signatures/). If the intent here is to have something like HMAC, you'll want to also revisit this.

paragonie-security reviewed (changes requested):
The current implementation needs work before it can be a general-purpose encryption component for all Symfony users.

fabpot commented:
@Nyholm What do we do here?

Nyholm commented:
Im gathering courage to make this mergable. I will not stall more than another month.

OskarStark commented:
Must target 6.1 then, right?

soatok commented on src/Symfony/Component/Encryption/Sodium/SodiumEncryption.php:
```diff
@@ -0,0 +1,96 @@
+            throw new DecryptionException(sprintf('Failed to decrypt message with algorithm "%s".', $algorithm), $exception);
+        }
+
+        if (false === $output) {
```
`!is_string()` would be a better condition for static analysis.

ro0NL commented:
s/encryption/encryptor/ everywhere? :sweat_smile:

fabpot commented:
Let's close for now
